### DISCUSSION TITLE
How to switch to workspace number N and how to move window to workspace number N

### DISCUSSION THREAD

**Discussion by marc-hg** [dalyIsaac/Whim#860] [Q&A] [answered]
Switching to workspace number 1, 2, 3.. is a feature i use a lot in i3, I press on winkey + 3 it switches to workspace 3. I can also do shift + winkey + 3 and it switches current window to window 3.

Is there any way to do this with on Whim?

**Comment by urob** [comment]
Yes, the default keybind for switching to workspace `N` is `Alt + Shift + N`. There's no default keybind for moving windows to a specific workspace but it can be added easily using the API. 

FYI: a list of all default keybinds can be found here: https://dalyisaac.github.io/Whim/docs/customize/commands.html#core-commands

**Reply by marc-hg** [reply]
Thanksss. About the API could you give me osme insight on how would I go about moving selected window to N workspace?

**Reply by urob** [reply]
You'll need a [custom command](https://dalyisaac.github.io/Whim/docs/customize/commands.html#custom-commands) for that. Take a look how [`ActivateWorkspaceAtIndex`](https://github.com/dalyIsaac/Whim/blob/e7710db921fd21bef84dd2627656575d77ede166/src/Whim/Commands/CoreCommands.cs#L262-L271) is defined [here](https://github.com/dalyIsaac/Whim/blob/e7710db921fd21bef84dd2627656575d77ede166/src/Whim/Commands/CoreCommands.cs#L315-L325). You'll want to precede the `context.Butler.Activate` call with a `context.Butler.MoveWindowToWorkspace`.

**Reply by marc-hg** [reply]
I added this method to CoreCommands.cs (similar to ActivateWorkspaceAtIndex) what do you think?

```cs
	/// <summary>
	/// This record represents a command to move the currently focused window to a specific workspace, determined by the given index.
	/// The index here is 1-based.
	/// </summary>
	private record MoveWindowToWorkspaceAtIndex(int Index)
	{
		public void Execute(IContext context)
		{
			IWorkspace[] workspaces = context.WorkspaceManager.ToArray();
			if (Index <= workspaces.Length)
			{
				IWorkspace currentWorkspace = context.WorkspaceManager.ActiveWorkspace;
				IWindow? currentWindow = context.WorkspaceManager.ActiveWorkspace.LastFocusedWindow;
				if (currentWindow != null)
				{
					context.Butler.MoveWindowToWorkspace(currentWorkspace, currentWindow);
				}
			}
		}
	}
```

If it's not too much ask could you give me a quick rundown on how'd you test something like this? I'm not very familiar with dotnet projects. I assume there's a dotnet run command and/or a dotnet build to build the exe?

**Reply by urob** [reply, marked as answer]
No need to recompile Whim. As alluded to above, you can use custom commands to implement this in user space. Here's a simplified version that moves the current window to workspace 1:

```cs
// Add this at the top of your csx file where the other using directives are
using System.Linq;

// Add the below inside the DoConfig function inside your csx file

// Create the custom command
context.CommandManager.Add(
    identifier: "move_window_to_workspace_1",
    title: "Move window to workspace 1",
    callback: () =>
    {
        int Index = 1;  // use a fixed index as proof of concept
        IWorkspace[] workspaces = context.WorkspaceManager.ToArray();
    if (Index <= workspaces.Length)
    {
        context.Butler.MoveWindowToWorkspace(workspaces[Index - 1]);
        context.Butler.Activate(workspaces[Index - 1]);
    }
    }
);

// Create an associated keybind for Win + Shift + 1
context.KeybindManager.SetKeybind("whim.custom.move_window_to_workspace_1", new Keybind(IKeybind.WinShift, VIRTUAL_KEY.VK_1));
```

Pasting the above into your `csx` file should do the trick. (This assumes you are running the latest version of Whim). To create bindings for the other workspaces, you can either copy and paste this or, more elegantly, follow the implementation of `ActivateIndexAtWorkspace` linked above.

**Reply by urob** [reply]
PS: Given that everything is in userspace, testing is simply a matter of restarting Whim (can be done via the CommandOps). If there's an error, the popup is usually pretty informative to what has gone wrong.

**Reply by marc-hg** [reply]
I tried that script but it didn't quite work. It doesn't throw an error or anything it just doesn't move the window

**Reply by urob** [reply]
There was a small typo in the last line, missing the `custom` namespace when setting the keybind. I fixed it above. The piece that changes is `whim.custom.move_window_to_workspace_1` instead of `whim.move_window_to_workspace_1` in the last line.

**Reply by marc-hg** [reply]
thanks a bunch that worked
